### ISSUE TITLE
[Phase 3] 統合テスト - Metal GPU Neural Network

### ISSUE THREAD

**Issue by Showhey798** [Showhey798/gradflow#18]
## 概要

Metal GPU で Neural Network を学習し、Phase 3 の統合テストを実施します。

## テスト項目

- [ ] Metal GPU でのニューラルネットワーク学習
- [ ] 勾配が正しく計算される
- [ ] Unified Memory の効率性を確認

## テストコード例

\`\`\`cpp
TEST(Phase3Integration, MetalGPUNeuralNetwork) {
    // データを Metal GPU に配置
    auto X = Variable<float>::randn({1000, 100}, true, DeviceType::Metal);
    auto W1 = Variable<float>::randn({100, 50}, true, DeviceType::Metal);
    auto b1 = Variable<float>::zeros({50}, true, DeviceType::Metal);
    auto W2 = Variable<float>::randn({50, 10}, true, DeviceType::Metal);
    auto b2 = Variable<float>::zeros({10}, true, DeviceType::Metal);
    
    // Forward
    auto h = relu(matmul(X, W1) + b1);
    auto y = matmul(h, W2) + b2;
    
    // Backward
    auto loss = y.sum();
    loss.backward();
    
    // 勾配が計算されていることを確認
    EXPECT_TRUE(W1.grad().device() == DeviceType::Metal);
    EXPECT_GT(W1.grad().abs().sum().data()[{0}], 0.0);
    
    // CPU に転送して結果を確認（Unified Memory で効率的）
    auto y_cpu = y.to(DeviceType::CPU);
    EXPECT_EQ(y_cpu.shape(), Shape({1000, 10}));
}
\`\`\`

## 完了基準

- [ ] Metal GPU での学習が成功
- [ ] CPU と Metal GPU で同じ結果
- [ ] パフォーマンスベンチマーク実施

## 関連ドキュメント

- [ROADMAP.md Phase 3 統合テスト](docs/ROADMAP.md)

## 期間

Week 4

## 依存タスク

- Depends on #17
